### DISCUSSION TITLE
Implement an ErrorListener

### DISCUSSION THREAD

**Discussion by madchicken** [rrevenantt/antlr4rust#85] [Q&A]
Hi, I am trying to implement an ErrorListener for my parser to collect syntax errors, but the function `add_error_listener` takes a boxed instance of my listener, preventing me from reusing it later in my code (to extract what I collected during the parse. phase, for example). Was it done on purpose or the method signature is wrong? Am I missing something here?

**Comment by pm100** [comment]
@madchicken got the same issue. What did you end up doing? 

I assume you have the same problem. I want to know after the parse pass that there was an error and stop, not run my visitor.

I notice that the DefaultErrorStrategy inserts ErrorNodes into the resulting tree, but I cannot work out how to detect them. If I print them via {:?} the error text shows but I can work out how to ask for the type.

I have to say that this is an amazingly hard tool to use (I used antlr a lot before in other languages, easy). Its totally not obvious how to build a non trivial , production ready product with it. Toying with giving up and using pest instead , but I have a working visitor implementation, I just cant et the error handling right

**Reply by madchicken** [reply]
@pm100 yes, I could eventually solve my issue. Please [look here for a response](https://stackoverflow.com/questions/77661616/rust-reuse-moved-struct-instance) (it's me responding to myself 🙃).
Regarding the rest of your comment, I agree with you and I think this project needs more attention and maintenance before it can be used in a project. I switched to pest too.
